images/bootstrap/pkg/service/cm_lock.go

```diff
@@ -0,0 +1,93 @@
+package service
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type ConfigMapLock struct {
+	kubernetes.Interface
+}
+
+func (c *ConfigMapLock) AcquireLock() bool {
+	cfg, err := c.CoreV1().ConfigMaps(namespace).Get(configmap, meta_v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		log.Warnf("bootstrap: configmap cassandra-topology configmap %s cannot be found...", configmap)
+		return false
+	}
+	if cfg.Annotations == nil {
+		cfg.Annotations = make(map[string]string)
+	}
+	cfg.Name = configmap
+	if cfg.Annotations[annotationLock] == "" {
+		cfg.Annotations[annotationLock] = pod
+		log.Infof("bootstrap: acquiring annotationLock of configmap %s/%s", namespace, configmap)
+		_, err = c.CoreV1().ConfigMaps(namespace).Update(cfg)
+		return err == nil
+	}
+	return false
+}
+
+func (c *ConfigMapLock) HasLock() bool {
+	cfg, err := c.CoreV1().ConfigMaps(namespace).Get(configmap, meta_v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		log.Warnf("bootstrap: configmap cassandra-topology configmap %s cannot be found...", configmap)
+		return false
+	}
+	if cfg.Annotations == nil {
+		return false
+	}
+	return cfg.Annotations[annotationLock] == pod
+}
+
+func (c *ConfigMapLock) ReleaseLock() bool {
+	if c.HasLock() {
+		cfg, err := c.CoreV1().ConfigMaps(namespace).Get(configmap, meta_v1.GetOptions{})
```

Comment:
the lock was implemented with `re-entrant mutex` lock pattern, so we discard any option of having a deadlock here. But I understand what you mean here and it can be simplified in some cases here. Let me take a stab here!

Reply:
I have optimized the lock logic altogether and this as part of the improvement